src/Server/TCPHandler.cpp

```diff
@@ -1151,7 +1154,20 @@ void TCPHandler::receiveHello()
     }
 
     session = makeSession();
-    session->authenticate(user, password, socket().peerAddress());
+    auto & client_info = session->getClientInfo();
+
+    /// Extract the last entry from comma separated list of forwarded_for addresses.
+    /// Only the last proxy can be trusted (if any).
+    Strings forwarded_addresses;
```

Comment:
Instead of doing this in different places, I'd suggest create a some field in `ClientInfo` to set trusted forwarded address there when `forwarded_for` is set.